### PR TITLE
Fix masp rewards estimate

### PR DESCRIPTION
**Pull request by grarco**
## Describe your changes

Various adjustments to the masp rewards estimation logic:

- Fixed the logic to correctly account for rewards accrued by the native token and by assets that have been shielded in the current epoch (for which no conversions are available yet but the conversion from the previous epoch can be used as an estimate)
- Fixed the logic to allow estimating rewards for assets that have been shielded before the last epoch
- Changed the API of the estimate function to require a `ValueSum` instead of a viewing key (this is to support a UI view with the possible rewards for different tokens, cc @ChrisHoltDesign, @mateuszjasiuk)
- The function now returns a `DenominatedAmount` instead of an `i128`
- Changed the name of the cli command to `estimate-shielding-rewards`
- Updated and improved tests

## Checklist before merging 
- [ ] If this PR has some consensus breaking changes, I added the corresponding `breaking::` labels
    - This will require 2 reviewers to approve the changes
- [ ] If this PR requires changes to the docs or specs, a corresponding PR is opened in the `namada-docs` repo
    - Relevant PR if applies: 
- [ ] If this PR affects services such as `namada-indexer` or `namada-masp-indexer`, a corresponding PR is opened in that repo
    - Relevant PR if applies:

### CODE REVIEW

File: crates/shielded_token/src/masp/shielded_wallet.rs [hunk 1]
```diff
@@ -713,95 +709,228 @@ pub trait ShieldedApi<U: ShieldedUtils + MaybeSend + MaybeSync>:
...
+                .0;
+            let mut est_conv = I128Sum::zero();
+            for ((_, mut asset_data), val) in decoded_conv.into_components() {
+                // We must upconvert the components of the conversion unless
+                // this conversion is for a non-native token and the component
+                // itself is the native token. In this case we should avoid
+                // upconverting it since conversions for non-native tokens must
+                // always refer to a specific epoch (MaspEpoch(0)) native token
+                if asset.token == native_token
+                    || asset_data.token != native_token
```
Comment (batconjurer) [inline crates/shielded_token/src/masp/shielded_wallet.rs:809]:
```suggestion
                if !(asset.token != native_token
                    && asset_data.token == native_token)
```

Comment (batconjurer) [reply crates/shielded_token/src/masp/shielded_wallet.rs:809]: This matches the comment more closely. I was confused until I wrote it out like this.

File: crates/shielded_token/src/masp/shielded_wallet.rs [hunk 2]
```diff
@@ -713,95 +709,228 @@ pub trait ShieldedApi<U: ShieldedUtils + MaybeSend + MaybeSync>:
...
+
+        // Extract only the native token balance for rewards. Depending on
+        // rewards being enabled for the native token or not, we'll find the
+        // balance we look for at either epoch 0 or the current epoch. If
+        // rewards are not enabled for the native token we might have
+        // balances at any epochs in between these two but these would be
+        // guaranteed to be the same in both the exchanged amounts we've
+        // computed here above, meaning they are irrelevant for the
+        // estimation of the next epoch rewards since they would cancel out
+        // anyway
```
Comment (batconjurer) [inline crates/shielded_token/src/masp/shielded_wallet.rs:842]:
```suggestion
        // Extract only the native token balance for rewards. Depending on
        // rewards being enabled for the native token or not, we'll find the
        // balance we look for at either epoch 0 or the current epoch. If
        // rewards are not enabled for the native token we might have
        // balances at any epochs in between these two. These would be
        // guaranteed to be the same in both the exchanged amounts we've
        // computed here above. This means they are irrelevant for the
        // estimation of the next epoch rewards since they would cancel out
        // anyway.
```

Comment (batconjurer) [reply crates/shielded_token/src/masp/shielded_wallet.rs:842]: nit: run on sentence

File: crates/shielded_token/src/masp/shielded_wallet.rs [hunk 3]
```diff
@@ -713,95 +709,228 @@ pub trait ShieldedApi<U: ShieldedUtils + MaybeSend + MaybeSync>:
...
+                        return Err(eyre!(
+                            "Found reward asset other than the native token"
+                        ));
+                    }
+                    match epoch {
+                        Some(ep) if ep == epoch_zero => (),
+                        _ => {
+                            return Err(eyre!(
+                                "Found reward asset with an epoch different \
+                                 than the current one"
```
Comment (batconjurer) [inline crates/shielded_token/src/masp/shielded_wallet.rs:920]: This error message doesn't seem correct.

File: crates/shielded_token/src/masp/shielded_wallet.rs [hunk 4]
```diff
@@ -2007,21 +2245,37 @@ mod test_shielded_wallet {
...
 
+                // If this flag is clear we test that the estimation logic can
+                // hanlde rewards for assets coming only from the current epoch,
+                // i.e. assets for which there are no conversions yet.
+                if shield_at_previous_epoch {
+                    let asset_data = AssetData {
+                        token: incentivized_token.clone(),
+                        denom: 0.into(),
+                        position: MaspDigitPos::Zero,
+                        epoch: Some(MaspEpoch::new(1)),
```
Comment (batconjurer) [inline crates/shielded_token/src/masp/shielded_wallet.rs:2270]: isn't the current epoch 2?

File: crates/shielded_token/src/masp/shielded_wallet.rs [hunk 5]
```diff
@@ -2032,11 +2286,184 @@ mod test_shielded_wallet {
...
+                    .await
+                    .unwrap()
+                    .unwrap_or_else(ValueSum::zero);
+                let rewards_est = wallet.estimate_next_epoch_rewards(&context, &balance)
+                    .await.expect("Test failed");
+                assert_eq!(
+                    rewards_est,
+                    DenominatedAmount::native(
+                        Amount::from_masp_denominated_i128(
+                            (1 + i128::from(shield_at_previous_epoch)) * reward_rate * i128::from(principal),
```
Comment (batconjurer) [inline crates/shielded_token/src/masp/shielded_wallet.rs:2303]: I'm confused about the `1+` here.

File: crates/shielded_token/src/masp/shielded_wallet.rs [hunk 6]
```diff
@@ -2032,11 +2286,184 @@ mod test_shielded_wallet {
...
+
+        /// In this test, we have a single incentivized token, the native one,
+        /// shielded at MaspEpoch(2), owned by the shielded wallet.
+        /// The amount of owned token is the parameter `principal`.
+        ///
+        /// We add a conversion from MaspEpoch(1) to MaspEpoch(2)
+        /// which issues `reward_rate` nam tokens for our incentivized token.
+        ///
+        /// Optionally, the test also shields the same `principal` amount at
+        /// MaspEpoch(1).
```
Comment (batconjurer) [inline crates/shielded_token/src/masp/shielded_wallet.rs:2318]: Again, maybe I'm confused about what's being tested, but shouldn't you shield at epoch 2 (the current epoch)?

File: crates/shielded_token/src/masp/shielded_wallet.rs [hunk 7]
```diff
@@ -2032,11 +2286,184 @@ mod test_shielded_wallet {
...
+                    .await
+                    .unwrap()
+                    .unwrap_or_else(ValueSum::zero);
+                let rewards_est = wallet.estimate_next_epoch_rewards(&context, &balance).await.expect("Test failed");
+                if shield_at_previous_epoch {
+                    assert_eq!(
+                        rewards_est,
+                        DenominatedAmount::native(
+                            Amount::from_masp_denominated_i128(
+                                (2 + reward_rate) * reward_rate * i128::from(principal),
```
Comment (batconjurer) [inline crates/shielded_token/src/masp/shielded_wallet.rs:2451]: Another mysterious `2+`. Maybe write out the formula's meaning in a comment.

File: crates/shielded_token/src/masp/shielded_wallet.rs [hunk 8]
```diff
@@ -2272,22 +2782,27 @@ mod test_shielded_wallet {
...
-                let principal2 = i128::from(principal2);
-                // reward from epoch 1->2 + reward from epoch 2->3 + reward from
-                // epoch 2->3
-                let expected_tok1_rewards =
-                    principal1 * tok1_reward_rate + principal1 + principal1;
-                // reward from epoch 2->3 + reward from epoch 2->3
-                let expected_tok2_rewards =
-                    principal2 * tok2_reward_rate + principal2 * tok2_reward_rate;
+                // The native token balances at epoch 3 and 4 are given by the shielded amounts times the rewards times the conversion for the native asset
+                let native_balance_at_3 = (i128::from(principal1) * (tok1_reward_rate + 1) + i128::from(principal2) * tok2_reward_rate) * 2;
```
Comment (batconjurer) [inline crates/shielded_token/src/masp/shielded_wallet.rs:2795]: Wait, maybe I've got it. Is this supposed to return the balance, i.e. the principal + rewards? If so, then these mysterious `x+` make sense.

**Review by batconjurer — CHANGES_REQUESTED** [review]

File: crates/apps_lib/src/client/rpc.rs [hunk 9]
```diff
@@ -371,19 +369,46 @@ pub async fn query_proposal_by_id<C: Client + Sync>(
...
 pub async fn query_rewards_estimate(
     context: &impl Namada,
-    args: args::QueryRewardsEstimate,
+    args: args::QueryShieldingRewardsEstimate,
 ) {
     let mut shielded = context.shielded_mut().await;
     let _ = shielded.load().await;
-    let rewards_estimate = shielded
-        .estimate_next_epoch_rewards(context, &args.owner.as_viewing_key())
+    let raw_balance = match shielded
```
[the remark this answers is not in the document]
Comment (grarco) [reply crates/apps_lib/src/client/rpc.rs:376]: To allow passing arbitrary balances to that function so that we can use it in the interface to present the user with an estimate of the rewards they can expect

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 1 above)
Comment (grarco) [reply crates/shielded_token/src/masp/shielded_wallet.rs:809]: You're right the comment is a bit misleading, I changed that though instead of modifying the if condition

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 4 above)
Comment (grarco) [reply crates/shielded_token/src/masp/shielded_wallet.rs:2271]: Yes but if `shielded_at_previous_epoch` is **clear** then we only shield at the current epoch (2), if it is set instead we also shield at the previous one, let me rephrase the comment

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 5 above)
Comment (grarco) [reply crates/shielded_token/src/masp/shielded_wallet.rs:2304]: So the expected rewards are just `principal * reward_rate` but if we also shielded at the previous epoch then we need to multiply this value by 2, hence the first addition. You think it would be cleaner with two different assertions selected by an `if` on `shield_at_previous_epoch`?

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 6 above)
Comment (grarco) [reply crates/shielded_token/src/masp/shielded_wallet.rs:2473]: Yes but optionally I also want to shield an additional `principal` amount at the previous epoch to check that the rewards estimate logic holds

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 7 above)
Comment (grarco) [reply crates/shielded_token/src/masp/shielded_wallet.rs:2606]: Same answer as before, only difference is that rewards for the native token compound, so we need to add 2 instead of 1 to the multiplier this time

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 8 above)
Comment (grarco) [reply crates/shielded_token/src/masp/shielded_wallet.rs:3081]: Only in this case, for the previous two cases no, it's supposed to return the estimated rewards

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 7 above)
Comment (grarco) [reply crates/shielded_token/src/masp/shielded_wallet.rs:2606]: Same as before, but in this case we should also account for the compounding of rewards when the native token itself accrues rewards, I've added a comment to the formula

File: crates/shielded_token/src/masp/shielded_wallet.rs [hunk 10]
```diff
@@ -713,95 +709,228 @@ pub trait ShieldedApi<U: ShieldedUtils + MaybeSend + MaybeSync>:
...
+            asset.redate_to_next_epoch();
+            let decoded_conv = self
+                .decode_sum(context.client(), conv.clone().into())
+                .await
+                .0;
+            let mut est_conv = I128Sum::zero();
+            for ((_, mut asset_data), val) in decoded_conv.into_components() {
+                // We must upconvert the components of the conversion unless
+                // this conversion is for a non-native token and the component
+                // itself is the native token. In this case we should avoid
```
Comment (batconjurer) [inline crates/shielded_token/src/masp/shielded_wallet.rs]:
```suggestion
                //  the component itself is the native token and this conversion
                //  is for a non-native token. In this case we should avoid
```

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 3 above)
Comment (batconjurer) [reply crates/shielded_token/src/masp/shielded_wallet.rs]: I feel like you left me hanging here. You checked if the epoch is zero and error otherwise.

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 8 above)
Comment (batconjurer) [reply crates/shielded_token/src/masp/shielded_wallet.rs:3093]: 🙈

**Review by batconjurer — CHANGES_REQUESTED** [review]
Just some nits left.

File: crates/shielded_token/src/masp/shielded_wallet.rs (hunk 3 above)
Comment (grarco) [reply crates/shielded_token/src/masp/shielded_wallet.rs]: So we extract the native token components from the result of the subtraction of the two balances. Then, to be able to sum all these components and return  a `DenominatedAmount`, we normalize them at `MaspEpoch(0)`. But we could pick any other epoch so long as it's a fixed one. The new error message is:

`Found reward asset with an epoch different than the 0th`

Should I change it to something else?

Comment (batconjurer) [reply crates/shielded_token/src/masp/shielded_wallet.rs]: I'm good with that

**Comment by brentstone** [comment]
@grarco what is the status here? Any chance this can be included in the next release?

**Comment by grarco** [comment]
> @grarco what is the status here? Any chance this can be included in the next release?

I've just rebased it to solve a conflict with main, I guess I need approval from @batconjurer

**Review by batconjurer — APPROVED** [review]
